src/cpp/src/sequence_group.hpp

```diff
@@ -649,7 +649,11 @@ class SequenceGroup  : public std::enable_shared_from_this<SequenceGroup> {
                 if (has_finished()) {
                     m_stream_window_size = 0;
                 }
+                // push empty output in case we won't stream generation res
                 if (generated_len <= (m_num_streamed_tokens + m_stream_window_size)) {
+                    if (has_finished()) {
+                        push_empty_outputs();
+                    }
                     return;
                 }
```

Comment:
There's still a path in `notify_handle` method that results in return without push - in such scenario call to `notify_handle` does not effectively notify the handle. I think we should stick to the principle that this method always calls one of the push methods so we are sure that if it's called, handle is always notified. I believe this way we reduce likelihood of bugs in the future.

Perhaps we can make some of the checks here outside this method and don't call it if conditions are not met?

Reply:
I think it won't be effective from performance perspective if we will push empty outputs per iteration comparing with once empty push per finished sequence.
Recently, we have implemented threading streaming, so there is conditional variable waiting for pushed outputs. If we will unlock thread per empty push, we won't just wait, we will try to stream empty generation

Reply:
We definitely don't want to push empty outputs over and over. The condition for such empty push is correct, I just don't like the idea that `notify_handle` can be called and don't push any outputs. 

I wonder if we could merge those conditions:
```
if (generated_len <= (m_num_streamed_tokens + m_stream_window_size) && has_finished())
	push_empty_outputs();
```
And check if `generated_len <= (m_num_streamed_tokens + m_stream_window_size) && !has_finished()` outside of this method (somewhere in sampler?) and if it's true then don't call notify at all. 

It's more code, but conceptually it feels "cleaner" in my opinion. 
But this change is functionally okay, so from that point of view I'm okay with that.

Reply:
@mzegla

The current implementation is optimal from the perf perspective as I mentioned in my previous comment:
```
                if (generated_len <= (m_num_streamed_tokens + m_stream_window_size)) {
                    if (has_finished()) {
                        push_empty_outputs();
                    }
                    return;
                }
```

So, from client side nothing will change if we will push empty output for every iteration with not streamed tokens (like start stop_string matching). But it will affect the threading based streaming, because we will unlock conditional variable and try to stream instead of waiting.